### PR TITLE
Add pq dist table support

### PR DESCRIPTION
**Pull request by gusye1234**
Milestones:

- [x] Add local state support for `hnswlib` 
- [x] Index queries' distance tables using local state
- [ ] Optimize searching precision for the complete PQ backend

### CODE REVIEW

File: annlite/core/index/hnsw/index.py [hunk 1]
```diff
@@ -36,8 +36,14 @@ def pre_processed(self: 'HnswIndex', x: np.ndarray, *args, **kwargs):
             elif not self._set_backend_pq:
                 self._index.loadPQ(self.pq_codec)
                 self._set_backend_pq = True
+            dtables = self.pq_codec.get_dist_mat(x)
             x = self.pq_codec.encode(x)
-        return f(self, x, *args, **kwargs)
+
+            assert dtables.dtype == 'float32'
+            assert dtables.flags['C_CONTIGUOUS']
```
Comment (numb3r3) [inline annlite/core/index/hnsw/index.py:43]: Please remove the two asserts, I cannot see the benifits.

File: annlite/core/index/hnsw/index.py [hunk 2]
```diff
@@ -36,8 +36,14 @@ def pre_processed(self: 'HnswIndex', x: np.ndarray, *args, **kwargs):
             elif not self._set_backend_pq:
                 self._index.loadPQ(self.pq_codec)
                 self._set_backend_pq = True
+            dtables = self.pq_codec.get_dist_mat(x)
```
Comment (numb3r3) [inline annlite/core/index/hnsw/index.py:39]: I would prefer to use `kwargs['dist_tables'] = self.pq_codec.get_dist_mat(x)` ?

**Review by numb3r3 — CHANGES_REQUESTED** [review]
And please rebase your PR to fix conflicts

File: annlite/core/index/hnsw/index.py (hunk 1 above)
Comment (gusye1234) [reply annlite/core/index/hnsw/index.py:43]: Here assertions are to make sure the memory layout is expected. Since the `dtable`'s buffer will then be read as the float array and understood as the row-major storage style.
If those two asserts can't be satisfied, undefined behaviours may happen.

File: include/hnswlib/space_ip.h [hunk 3]
```diff
@@ -68,7 +70,8 @@ static float InnerProductSIMD4Ext(const void *pVect1v, const void *pVect2v,
 #elif defined(USE_SSE)
 
 static float InnerProductSIMD4Ext(const void *pVect1v, const void *pVect2v,
-                                  const void *qty_ptr) {
+                                  const void *qty_ptr, ,
```
Comment (numb3r3) [inline include/hnswlib/space_ip.h:73]:
```suggestion
                                  const void *qty_ptr,
```

**Review by numb3r3 — CHANGES_REQUESTED** [review]

**Comment by numb3r3** [comment]
@gusye1234 Please reset your last commit to pass lint check

**Review by numb3r3 — APPROVED** [review]
LGTM, great job.

**Comment by patelprateek** [comment]
@numb3r3 @gusye1234 , can you please provide some more context on this PR about what exactly is local state and why we need it , IIRC pq was supported codec in hnsw . thanks in advance , 
trying to debug some segfaults i have been running into recently .

**Comment by gusye1234** [comment]
Hi @patelprateek , 
by default, the functionality in this PR is disabled
https://github.com/jina-ai/annlite/blob/02857ec4103e32139f695adac4f7a40c4e65c67a/annlite/index.py#L62
So if you didn't explicitly enable it and the segfaults was at the runtime, not the compiling time, then probably this PR is not what you looking for.

> IIRC pq was supported codec in hnsw 

The pq `annlite` supported before was `PQIndex`, which is a kind of strategies for linear search. This PR supports the pq inside `HNSW`. The local state here is the context that pq needs to perform the distance table loop-up inside the `hnswlib`
https://github.com/jina-ai/annlite/blob/02857ec4103e32139f695adac4f7a40c4e65c67a/include/hnswlib/hnswlib.h#L34-L36

**Comment by patelprateek** [comment]
@gusye1234 : I thought the support for pq in hnsw was added in this PR : https://github.com/jina-ai/annlite/pull/122 by you.
So i thought we already support PQ quantization in hnsw . But now with this current PR perhaps I am missing  some context

**Comment by gusye1234** [comment]
@patelprateek , yeah...
#122 did support a naive PQ algorithm, which is less inaccurate, since a global distance table was used.
This PQ fixed that problem.

**Comment by patelprateek** [comment]
@gusye1234 : thanks for the prompt reply and adding this support . 
For my own curiosity and learning (since i am somewhat new to quantization) can you provide me some pointers or docs to : 
- what was the initial global PQ support , why is it inaccurate
- What was the new improved algorithm. 
From couple of paper i read on PQ i just knew there are 2 distance symmetric and asymmetric , but from PR it wasnt very clear to me , so any pointers or rough idea will help

**Comment by gusye1234** [comment]
Suppose the original space is a 128-dim, and we want to quantize it to  8-dim. For PQ, that means we need to create 8 subspaces and each subspace should be in charge of 16-dim of the original space. We also assume for each subspace we  use total of 256 possible code(0,1,2,...,255), and we can have a 16-dim center for each code.


Normally, a PQ should have a function to compute the distance table for an original 128-dim vector, say `a`. The distance table is specified for `a` with shape `(8, 256)`, which contains the distances of subvectors of `a` against the 256 centers for total of 8 subspaces.
> what was the initial global PQ support , why is it inaccurate

In the first version #122, we hit a rock that the abstract of `hnswlib`(current implementation for HNSW) really didn't support passing the customized(local) data for such vector `a`. It requires you to set up all the data globally. So we decided to firstly run through the pipeline with a simplified PQ. That is, instead of computing the distance table for each vector, a global distance table is computed with shape `(8, 256, 256)`. That table contains the distances between cluster centers. So we can take two 8-dim quantized vectors, and perform a loop-up on this table. 

In other words, the "local" distance table(`(8, 256)`) for vector `a` contains the distances between `a` and other quantized codes. The "global" distance table`(8, 256, 256)` in #122 contains distances of the quantized codes. 
The reason why #122 is inaccurate was instead of using the original `a`, we used the quantized `a` which certainly lost the information,
> What was the new improved algorithm?

We didn't "improve". #122 is a fix-up and not actually PQ. We then changed the abstracts and the interfaces of `hnswlib` in this PR to support the real PQ. That means, now we can pass the `(8, 256)` distance table for vector `a` to compute the more accurate distances.

Sorry for the text flooding and hope this could help you to understand the diffs between #122 and this PR

**Comment by patelprateek** [comment]
@gusye1234 : thanks for the detailed explanation. Yes, when i took a quick look at #122 i could not find how we were computing this distance on per query vector basis. 

What you explained here makes perfect sense and in fact is exactly what i was referring to in the paper.
The initial implementation : distance between quantized query and item vectors is called symmetric distance computation and the improvement on per query basis where we dont quantize query vector and in-fact use local state to compute a more accurate distance is called asymmetric distance computation , which the original paper also mentioned is a more accurate .

**Comment by patelprateek** [comment]
@gusye1234 I wonder if we should support both methods ? Curios on your thoughts about any performance implication as well during runtime when creating local state for each query vs improvement in recall

**Comment by matchyc** [comment]
Any benchmark?
